File: packages/modules/lodash/core/swift/src/string.swift
import Foundation

/**
 * Converts string to camel case.
 */
public func camelCase(_ str: String?) -> String {
    guard let str = str else { return "" }
    let wordList = words(str)
    return wordList.enumerated().map { index, word in
        if index == 0 {
            return word.lowercased()
        } else {
            return word.lowercased().capitalizingFirstLetter()
        }
    }.joined()
}

/**
 * Converts the first character of string to upper case and the remaining to lower case.
 */
public func capitalize(_ str: String?) -> String {
    guard let str = str else { return "" }
    return str.lowercased().capitalizingFirstLetter()
}

/**
 * Checks if string ends with the given target string.
 */
public func endsWith(_ str: String?, _ target: String?, _ position: Int? = nil) -> Bool {
    guard let str = str, let target = target else { return false }
    let pos = position ?? str.count
    let end = str.index(str.startIndex, offsetBy: clamp(pos, 0, str.count))
    return str[..<end].hasSuffix(target)
}

/**
 * Converts the characters "&", "<", ">", '"', and "'" in string to their corresponding HTML entities.
 */
public func escape(_ str: String?) -> String {
    guard let str = str else { return "" }
    return str.replacingOccurrences(of: "&", with: "&amp;")
        .replacingOccurrences(of: "<", with: "&lt;")
        .replacingOccurrences(of: ">", with: "&gt;")
        .replacingOccurrences(of: "\"", with: "&quot;")
        .replacingOccurrences(of: "'", with: "&#39;")
}

/**
 * Converts string to kebab case.
 */
public func kebabCase(_ str: String?) -> String {
    guard let str = str else { return "" }
    return words(str).map { $0.lowercased() }.joined(separator: "-")
}

/**
 * Converts the first character of string to lower case.
 */
public func lowerFirst(_ str: String?) -> String {
    guard let str = str else { return "" }
    return str.prefix(1).lowercased() + str.dropFirst()
}

/**
 * Pads string on the left and right sides if it's shorter than length.
 */
public func pad(_ str: String?, _ length: Int, _ chars: String = " ") -> String {
    guard let str = str else { return "" }
    if str.count >= length { return str }
    let totalPad = length - str.count
    let leftPad = totalPad / 2
    return padStart(str, str.count + leftPad, chars).let { padEnd($0, length, chars) }
}

/**
 * Pads string on the right side if it's shorter than length.
 */
public func padEnd(_ str: String?, _ length: Int, _ chars: String = " ") -> String {
    guard let str = str else { return "" }
    if str.count >= length { return str }
    let diff = length - str.count
    let padContent = String(repeating: chars, count: (diff / chars.count) + 1).prefix(diff)
    return str + padContent
}

/**
 * Pads string on the left side if it's shorter than length.
 */
public func padStart(_ str: String?, _ length: Int, _ chars: String = " ") -> String {
    guard let str = str else { return "" }
    if str.count >= length { return str }
    let diff = length - str.count
    let padContent = String(repeating: chars, count: (diff / chars.count) + 1).prefix(diff)
    return padContent + str
}

/**
 * Repeats the given string n times.
 */
public func repeatString(_ str: String?, _ n: Int = 1) -> String {
    guard let str = str, n > 0 else { return "" }
    return String(repeating: str, count: n)
}

/**
 * Replaces matches for pattern in string with replacement.
 */
public func replace(_ str: String?, _ pattern: String, _ replacement: String) -> String {
    guard let str = str else { return "" }
    return str.replacingOccurrences(of: pattern, with: replacement)
}

/**
 * Converts string to snake case.
 */
public func snakeCase(_ str: String?) -> String {
    guard let str = str else { return "" }
    return words(str).map { $0.lowercased() }.joined(separator: "_")
}

/**
 * Splits string by separator.
 */
public func split(_ str: String?, _ separator: String?, _ limit: Int? = nil) -> [String] {
    guard let str = str else { return [] }
    guard let separator = separator else { return [str] }
    let parts = str.components(separatedBy: separator)
    if let limit = limit {
        return Array(parts.prefix(limit))
    }
    return parts
}

/**
 * Converts string to start case.
 */
public func startCase(_ str: String?) -> String {
    guard let str = str else { return "" }
    return words(str).map { $0.lowercased().capitalizingFirstLetter() }.joined(separator: " ")
}

/**
 * Checks if string starts with the given target string.
 */
public func startsWith(_ str: String?, _ target: String?, _ position: Int = 0) -> Bool {
    guard let str = str, let target = target else { return false }
    let start = str.index(str.startIndex, offsetBy: clamp(position, 0, str.count))
    return str[start...].hasPrefix(target)
}

/**
 * Converts string to lower case.
 */
public func toLower(_ str: String?) -> String {
    return str?.lowercased() ?? ""
}

/**
 * Converts string to upper case.
 */
public func toUpper(_ str: String?) -> String {
    return str?.uppercased() ?? ""
}

/**
 * Truncates string if it's longer than the given maximum string length.
 */
public func truncate(_ str: String?, _ length: Int = 30, _ omission: String = "...") -> String {
    guard let str = str else { return "" }
    if str.count <= length { return str }
    return str.prefix(length - omission.count) + omission
}

/**
 * The inverse of _.escape; this method converts the HTML entities &amp;, &lt;, &gt;, &quot;, and &#39; in string to their corresponding characters.
 */
public func unescape(_ str: String?) -> String {
    guard let str = str else { return "" }
    return str.replacingOccurrences(of: "&amp;", with: "&")
        .replacingOccurrences(of: "&lt;", with: "<")
        .replacingOccurrences(of: "&gt;", with: ">")
        .replacingOccurrences(of: "&quot;", with: "\"")
        .replacingOccurrences(of: "&#39;", with: "'")
}

/**
 * Converts the first character of string to upper case.
 */
public func upperFirst(_ str: String?) -> String {
    guard let str = str else { return "" }
    return str.prefix(1).uppercased() + str.dropFirst()
}

/**
 * Splits string into an array of its words.
 */
public func words(_ str: String?) -> [String] {
    guard let str = str else { return [] }
    let regex = try! NSRegularExpression(pattern: "[a-zA-Z0-9]+")
    let nsString = str as NSString
    let results = regex.matches(in: str, range: NSRange(location: 0, length: nsString.length))
    return results.map { nsString.substring(with: $0.range) }
}

private extension String {
    func capitalizingFirstLetter() -> String {
        return prefix(1).uppercased() + dropFirst()
    }
}

private extension NSObjectProtocol {
    func `let`<R>(_ block: (Self) -> R) -> R {
        return block(self)
    }
}

private func clamp<T: Comparable>(_ value: T, _ lower: T, _ upper: T) -> T {
    return min(max(value, lower), upper)
}

File: packages/modules/lodash/core/typescript/src/string.ts
/* eslint-disable @typescript-eslint/no-explicit-any */

function wordsCase(string: string): string[] {
  return (
    string.match(/[A-Z]{2,}(?=[A-Z][a-z]+|[0-9]|$)|\d+|[A-Z]?[a-z]+|[A-Z]+/g) ||
    []
  );
}

export function camelCase(string: string = ''): string {
  const w = wordsCase(string);
  if (w.length === 0) return '';
  return (
    w[0]!.toLowerCase() +
    w
      .slice(1)
      .map((w) => w[0]!.toUpperCase() + w.slice(1).toLowerCase())
      .join('')
  );
}

export function capitalize(string: string = ''): string {
  if (!string) return '';
  return string[0]!.toUpperCase() + string.slice(1).toLowerCase();
}

export function deburr(string: string = ''): string {
  return string.normalize('NFD').replace(/[\u0300-\u036f]/g, '');
}

export function endsWith(
  string: string = '',
  target: string,
  position?: number
): boolean {
  const pos = position === undefined ? string.length : position;
  return string.slice(0, pos).slice(-target.length) === target;
}

export function escape(string: string = ''): string {
  const htmlEscapes: Record<string, string> = {
    '&': '&amp;',
    '<': '&lt;',
    '>': '&gt;',
    '"': '&quot;',
    "'": '&#39;',
  };
  return string.replace(/[&<>"']/g, (char) => htmlEscapes[char] || char);
}

export function escapeRegExp(string: string = ''): string {
  return string.replace(/[.*+?^${}()|[\]\\]/g, '\\$&');
}

export function kebabCase(string: string = ''): string {
  return wordsCase(string)
    .map((w) => w.toLowerCase())
    .join('-');
}

export function lowerCase(string: string = ''): string {
  return wordsCase(string)
    .map((w) => w.toLowerCase())
    .join(' ');
}

export function lowerFirst(string: string = ''): string {
  if (!string) return '';
  return string[0]!.toLowerCase() + string.slice(1);
}

export function pad(
  string: string = '',
  length: number = 0,
  chars: string = ' '
): string {
  if (!string && length <= 0) return '';
  const totalPad = length - string.length;
  if (totalPad <= 0) return string;
  const padStart = Math.floor(totalPad / 2);
  const padEnd = totalPad - padStart;
  return createPadding(padStart, chars) + string + createPadding(padEnd, chars);
}

export function padEnd(
  string: string = '',
  length: number = 0,
  chars: string = ' '
): string {
  const totalPad = length - string.length;
  if (totalPad <= 0) return string;
  return string + createPadding(totalPad, chars);
}

export function padStart(
  string: string = '',
  length: number = 0,
  chars: string = ' '
): string {
  const totalPad = length - string.length;
  if (totalPad <= 0) return string;
  return createPadding(totalPad, chars) + string;
}

function createPadding(length: number, chars: string): string {
  if (chars === ' ') return ' '.repeat(length);
  return chars.repeat(Math.ceil(length / chars.length)).slice(0, length);
}

export function parseInt(string: string, radix: number = 10): number {
  return Number.parseInt(string, radix);
}

export function repeat(string: string = '', n: number = 1): string {
  return string.repeat(n);
}

export function replace(
  string: string = '',
  pattern: string | RegExp,
  replacement: string | ((substring: string, ...args: any[]) => string)
): string {
  return string.replace(pattern, replacement as any);
}

export function snakeCase(string: string = ''): string {
  return wordsCase(string)
    .map((w) => w.toLowerCase())
    .join('_');
}

export function split(
  string: string = '',
  separator: string | RegExp,
  limit?: number
): string[] {
  return string.split(separator, limit);
}

export function startCase(string: string = ''): string {
  return wordsCase(string)
    .map((w) => w[0]!.toUpperCase() + w.slice(1).toLowerCase())
    .join(' ');
}

export function startsWith(
  string: string = '',
  target: string,
  position: number = 0
): boolean {
  return string.slice(position, position + target.length) === target;
}

export function toLower(string: string = ''): string {
  return string.toLowerCase();
}

export function toUpper(string: string = ''): string {
  return string.toUpperCase();
}

export function trim(string: string = '', chars?: string): string {
  if (!chars) return string.trim();
  const pattern = new RegExp(
    `^[${escapeRegExp(chars)}]+|[${escapeRegExp(chars)}]+$`,
    'g'
  );
  return string.replace(pattern, '');
}

export function trimEnd(string: string = '', chars?: string): string {
  if (!chars) return string.trimEnd();
  const pattern = new RegExp(`[${escapeRegExp(chars)}]+$`, 'g');
  return string.replace(pattern, '');
}

export function trimStart(string: string = '', chars?: string): string {
  if (!chars) return string.trimStart();
  const pattern = new RegExp(`^[${escapeRegExp(chars)}]+`, 'g');
  return string.replace(pattern, '');
}

export function truncate(
  string: string = '',
  options: {
    length?: number;
    omission?: string;
    separator?: string | RegExp;
  } = {}
): string {
  const { length = 30, omission = '...', separator } = options;
  if (string.length <= length) return string;
  let result = string.slice(0, length - omission.length);
  if (separator) {
    if (typeof separator === 'string') {
      const lastIdx = result.lastIndexOf(separator);
      if (lastIdx > 0) result = result.slice(0, lastIdx);
    } else {
      const match = result.match(separator);
      if (match && match.index !== undefined && match.index > 0) {
        result = result.slice(0, match.index);
      }
    }
  }
  return result + omission;
}

export function unescape(string: string = ''): string {
  const htmlUnescapes: Record<string, string> = {
    '&amp;': '&',
    '&lt;': '<',
    '&gt;': '>',
    '&quot;': '"',
    '&#39;': "'",
    '&#x27;': "'",
  };
  return string.replace(
    /&(?:amp|lt|gt|quot|#39|#x27);/g,
    (entity) => htmlUnescapes[entity] || entity
  );
}

export function upperCase(string: string = ''): string {
  return wordsCase(string)
    .map((w) => w.toUpperCase())
    .join(' ');
}

export function upperFirst(string: string = ''): string {
  if (!string) return '';
  return string[0]!.toUpperCase() + string.slice(1);
}

export function words(string: string = '', pattern?: RegExp): string[] {
  if (pattern) {
    return string.match(pattern) || [];
  }
  return wordsCase(string);
}
